prompt2model/model_trainer/peft_trainer.py

```diff
@@ -0,0 +1,243 @@
+"""A trainer class for fine-tuning a model using QLora."""
+
+import gc
+import os
+
+import datasets
+import torch
+import transformers
+
+# import wandb
+from accelerate import Accelerator, FullyShardedDataParallelPlugin
+from peft import LoraConfig, PeftModel, get_peft_model, prepare_model_for_kbit_training
+from torch.distributed.fsdp.fully_sharded_data_parallel import (
+    FullOptimStateDictConfig,
+    FullStateDictConfig,
+)
+from transformers import AutoModelForCausalLM, AutoTokenizer, BitsAndBytesConfig
+
+fsdp_plugin = FullyShardedDataParallelPlugin(
+    state_dict_config=FullStateDictConfig(offload_to_cpu=True, rank0_only=False),
+    optim_state_dict_config=FullOptimStateDictConfig(
+        offload_to_cpu=True, rank0_only=False
+    ),
+)
+
+accelerator = Accelerator(fsdp_plugin=fsdp_plugin)
+
+
+class EvalAccuracyCallback(transformers.TrainerCallback):
+    """A callback to evaluate the model on the validation set and log the accuracy."""
+
+    def __init__(self, val_data: datasets.Dataset, eval_tokenizer) -> None:
+        """Initialize the callback with the validation data and tokenizer."""
+        range_len = min(100, len(val_data))
+        self.val_data = val_data.shuffle(seed=42).select(range(range_len))
+        self.eval_tokenizer = eval_tokenizer
+        self.device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+
+    def on_evaluate(self, args, state, control, **kwargs):
+        """Evaluate the model on the validation set and log the accuracy."""
```

Comment:
Add types and description of args.

Also, can we just replace `kwargs` with `model` and `metrics`? That will make things more robust.